File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> words(wordDict.begin(), wordDict.end());
        unordered_set<int> seen;
        queue<int> queue{{0}};
        
        while (!queue.empty()){
            int start = queue.front();
            queue.pop();
            if (start == s.size()) return true;
            
            for (int end = start + 1; end <= s.size(); ++end){
                if (seen.find(end) != seen.end()) continue;
                if (words.find(s.substr(start, end - start)) != words.end()){
                    queue.push(end);
                    seen.insert(end);
                }
            }
        }
        
        return false;
    }
};

File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        
        @lru_cache(maxsize = None)
        def dp(i: int) -> bool:
            if i < 0:
                return True
            
            for word in wordDict:
                if s[i - len(word) + 1: i + 1] == word and dp(i - len(word)):
                    return True
            
            return False
        n = len(s)
        return dp(n - 1)
        